rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -8,17 +8,258 @@ use crate::{
 };
 use once_cell::sync::Lazy;
 use pest::pratt_parser::{Assoc, Op, PrattParser};
-use std::sync::Arc;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+};
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
 pub enum Type {
+    NoValue,
     Atom(TypeAtom),
     Union(Arc<(Type, Type)>),
     Intersection(Arc<(Type, Type)>),
     Array(Arc<Type>),
     Dict(Arc<Type>),
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) struct CollapseError(Vec<String>);
+
+impl CollapseError {
+    fn join(errs: Vec<CollapseError>) -> CollapseError {
+        CollapseError(errs.into_iter().flat_map(|x| x.0).collect())
+    }
+}
+
+impl ToString for CollapseError {
+    fn to_string(&self) -> String {
+        self.0.clone().join(". ")
+    }
+}
+
+impl Type {
+    /// Build union out of DoubleEndedIterator of types
```

Comment:
Reading this and looking at the method I am left just as wise about it. Is it just creating a `Union` of all types on the iterator?

Reply:
yep. you said it better than the comment tho. changing the comment